pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/MailboxReceiveOperator.java

```diff
@@ -46,41 +49,56 @@ public String toExplainString() {
 
   @Override
   protected TransferableBlock getNextBlock() {
-    if (_errorBlock != null) {
-      return _errorBlock;
-    }
-    if (System.currentTimeMillis() > _context.getDeadlineMs()) {
-      _errorBlock = TransferableBlockUtils.getErrorTransferableBlock(QueryException.EXECUTION_TIMEOUT_ERROR);
-      return _errorBlock;
-    }
+    while (!_mailboxes.isEmpty()) {
+      if (_errorBlock != null) {
+        return _errorBlock;
+      }
+      if (System.currentTimeMillis() > _context.getDeadlineMs()) {
+        _errorBlock = TransferableBlockUtils.getErrorTransferableBlock(QueryException.EXECUTION_TIMEOUT_ERROR);
+        return _errorBlock;
+      }
 
-    // Poll from every mailbox in round-robin fashion:
-    // - Return the first content block
-    // - If no content block found but there are mailboxes not finished, return no-op block
-    // - If all content blocks are already returned, return end-of-stream block
-    int numMailboxes = _mailboxes.size();
-    for (int i = 0; i < numMailboxes; i++) {
-      ReceivingMailbox mailbox = _mailboxes.remove();
-      TransferableBlock block = mailbox.poll();
+      // Poll from every mailbox in round-robin fashion:
+      // - Return the first content block
+      // - If no content block found but there are mailboxes not finished, return no-op block
+      // - If all content blocks are already returned, return end-of-stream block
+      LOGGER.trace("==[RECEIVE]== Enter getNextBlock from: " + _context.getId() + " mailboxSize: " + _mailboxes.size());
```

Comment:
I guess this code is going to be executed in the hotpath. Take care here, because we are using the eager logging message. Therefore we are going to be creating and concatenating tons of strings even when TRACE is disabled. We should use the `{}` syntax or even better, do the explicit:

```
      if (LOGGER.isTraceEnabled()) {
        LOGGER.trace("==[RECEIVE]== Enter getNextBlock from: " + _context.getId() + " mailboxSize: " + _mailboxes.size());
      }
```

It looks like a very small thing, but I remember fixing a performance bug like that last year where we were spending a significant amount of time creating unused messages while applying segment pruning.

Reply:
BTW, the `if (LOGGER.isTraceEnabled())` code is uglier to read, but it is faster than `{}` because the later uses a vararg argument, which means Java creates an array under the hood.